Changes to Enable SSH for Repository hosting

itwasntandy opened the pull request:
We use Git over SSH for repository hosting. Adding the ability for
this cookbook to be used for configuring the phab host for this.
Runs an additional SSH Daemon on port 617 (configurable) for this
purpose.

Test cases added to ensure the daemon is installed and configured.

N.B. This is only for Ubuntu 14.04 or newer, due to need for
AuthorizedKeysCommand which came in OpenSSH 6.2.

Functionality is disabled on Ubuntu 12.04, and test cases are skipped
in that case.

itwasntandy committed:
Changes to Enable SSH for Repository hosting

We use Git over SSH for repository hosting. Adding the ability for
this cookbook to be used for configuring the phab host for this.
Runs an additional SSH Daemon on port 617 (configurable) for this
purpose.

Test cases added to ensure the daemon is installed and configured.

N.B. This is only for Ubuntu 14.04 or newer, due to need for
AuthorizedKeysCommand which came in OpenSSH 6.2.

Functionality is disabled on Ubuntu 12.04, and test cases are skipped
in that case.

kimtore commented on attributes/repo_hosting.rb:
```diff
@@ -0,0 +1,19 @@
+default['phabricator']['ssh_vcs_port'] = '617'
+
+# User for Source code hosting
+default['phabricator']['vcsuser'] = 'git'
```
I suggest that the VCS functions are namespaced into e.g. `['phabricator']['vcs_hosting']` (or, preferably, something a little more elegant)

itwasntandy replied:
Good idea, thanks

kimtore commented on recipes/repo_hosting.rb:
```diff
@@ -0,0 +1,107 @@
+
+user node['phabricator']['vcsuser'] do
+    comment 'VCS User'
+    home File.join('/home', node['phabricator']['vcsuser'])
```
The home directory must definitely be configured in node attributes, instead of auto-generated.

itwasntandy replied:
fair, will fix

kimtore commented on recipes/repo_hosting.rb:
```diff
@@ -0,0 +1,107 @@
+end
+
+# enable /etc/sudoers.d directory to enable the sudoer provider to work
+node.override[:authorization][:sudo][:include_sudoers_d] = true
```
Please use `['string']` syntax instead of `[:symbol]`.

itwasntandy replied:
eep, yes I don't know why I used symbols there. will fix.

kimtore commented on recipes/repo_hosting.rb:
```diff
@@ -0,0 +1,107 @@
+#
+# Cookbook Name:: phabricator
+# Recipe:: repo_hosting
```
Actually, I'd suggest not to use abbreviations unless they are really long (like VCS). `repo` could be replaced with `repository`.

kimtore commented:
Thank you very much for the patch. I'll look a bit further into it, possibly tomorrow, or at least when I have free time :+1:

kimtore commented:
Sorry for spamming with comments. Can you run your patch through Foodcritic to eliminate all linting errors? (with the exception that spaces should be set to 4, not 2)

itwasntandy commented:
Sure thing.. although I noted you're also using double quotes throughout the cookbook not single..  and are using the old style ruby hash syntax.

Would you like those fixed up, or would it be ok to commit a .rubocop.yml to set these preferences?

kimtore commented:
Ah, don't worry about those. I should fix these in a separate commit, thanks.

itwasntandy committed:
Tidying up as per code review

Making naming more consistent and breaking out the vcs hosting over ssh into a
separate attribute namespace under `node['phabricator']['vcs_ssh']`

Fixing some linter errors too, and deleting an unused template.

itwasntandy commented:
I pushed changes to this last week - are you happy with them? or would you like further change?

kimtore commented:
Sorry, been very busy at work so haven't gotten any time to look at it. I'll check it out tomorrow.

kimtore commented on attributes/vcs_ssh_hosting.rb:
```diff
@@ -0,0 +1,28 @@
+append_paths << '/usr/local/bin'
+append_paths << '/usr/lib/git-core'
+
+default['phabricator']['config']['environment.append-paths'] = %('#{append_paths}')
```
I'm curious to why this attribute is conofigured in `vcs_ssh_hosting.rb` and not `default.rb`? It looks like it belongs in the latter file.

kimtore commented on recipes/default.rb:
```diff
@@ -187,3 +187,12 @@
+# Setup ssh repo hosting if we want it!
+# Only applicable on 14.04 or newer, due to need for AuthorizedKeysCommand which came in
+# OpenSSH 6.2.
+if node['platform_version'] >= '14.04'
```
Setting `['phabricator']['vcs_ssh']['hosting_enabled']` to `true` on a system other than (Ubuntu) 14.04 will fail silently. It will be better to fail fast, throwing an exception in the `vcs_ssh_hosting` recipe if the system is unsupported.

Furthermore, what do you think about actually _not_ including the `vcs_ssh_hosting` recipe here, but instead including it explicitly in the run list if wanted? Seems easier to grasp what is actually happening.

kimtore commented on recipes/vcs_ssh_hosting.rb:
```diff
@@ -0,0 +1,110 @@
+end
+
+# enable /etc/sudoers.d directory to enable the sudoer provider to work
+node.override['authorization']['sudo']['include_sudoers_d'] = true
```
Why `override` and not `default`? Recipes have precedence over attribute files.

kimtore commented on recipes/vcs_ssh_hosting.rb:
```diff
@@ -0,0 +1,110 @@
+end
+
+# Install Git, Mercurial, SVN
+%w(git subversion mercurial).each do |pkg|
```
Good idea, but is it not the responsibility of the main recipe? IMO, in any case, it is a entirely different issue and should be committed separately.

kimtore commented on templates/default/vcs_ssh/sshd_config.erb:
```diff
@@ -0,0 +1,12 @@
+PermitRootLogin no
+AllowAgentForwarding no
+AllowTcpForwarding no
+PrintMotd no
```
Would be nice to have `X11Forwarding no`, even though it is the default, to be explicit about it.
